Author: Dragon-Chicken
Date: 2026-01-08
Description: Adds customisable gaps for when there's a single tiled window
Commit Version: f5d9381 (update logo)

Adds customisable gaps when there's a single tiled window in the workspace.
Two new configuration options 'single_horizontal_gap' and 'single_vertical_gap', which can be edited in the sxwmrc file.
Both options default 'gaps'.

---

diff --git a/src/defs.h b/src/defs.h
index ac8f74f..3da5441 100644
--- a/src/defs.h
+++ b/src/defs.h
@@ -82,6 +82,8 @@ typedef struct Client {
 typedef struct {
 	int modkey;
 	int gaps;
+	int single_horizontal_gap;
+	int single_vertical_gap;
 	int border_width;
 	long border_foc_col;
 	long border_ufoc_col;
diff --git a/src/parser.c b/src/parser.c
index e50db00..8e0d47b 100644
--- a/src/parser.c
+++ b/src/parser.c
@@ -376,6 +376,10 @@ int parser(Config *cfg)
 			cfg->border_foc_col = parse_col(rest);
 		else if (!strcmp(key, "gaps"))
 			cfg->gaps = atoi(rest);
+		else if (!strcmp(key, "single_horizontal_gap"))
+			cfg->single_horizontal_gap = atoi(rest);
+		else if (!strcmp(key, "single_vertical_gap"))
+			cfg->single_vertical_gap = atoi(rest);
 		else if (!strcmp(key, "master_width")) {
 			float mf = (float)atoi(rest) / 100.0f;
 			for (int i = 0; i < MAX_MONITORS; i++)
diff --git a/src/sxwm.c b/src/sxwm.c
index e5cd9d4..49c55b5 100644
--- a/src/sxwm.c
+++ b/src/sxwm.c
@@ -1546,6 +1546,8 @@ void init_defaults(void)
 {
 	user_config.modkey = Mod4Mask;
 	user_config.gaps = 10;
+	user_config.single_horizontal_gap = user_config.gaps;
+	user_config.single_vertical_gap = user_config.gaps;
 	user_config.border_width = 1;
 	user_config.border_foc_col = parse_col("#c0cbff");
 	user_config.border_ufoc_col = parse_col("#555555");
@@ -2605,6 +2607,44 @@ void tile(void)
 	if (total == 0)
 		return;
 
+	/* gaps when only a single window is tiled */
+	if (total == 1 && !monocle) {
+		for (Client *c = head; c; c = c->next) {
+
+			/* skip if not a tiled window */
+			if (!c->mapped || c->fullscreen || c->floating)
+				continue;
+
+			int border_width = user_config.border_width;
+
+			int single_horizontal_gap = user_config.single_horizontal_gap;
+			int single_vertical_gap = user_config.single_vertical_gap;
+
+			int mon = c->mon;
+			int x = mons[mon].x + mons[mon].reserve_left + single_horizontal_gap;
+			int y = mons[mon].y + mons[mon].reserve_top + single_vertical_gap;
+			int w = mons[mon].w - mons[mon].reserve_left - mons[mon].reserve_right - 2 * single_horizontal_gap;
+			int h = mons[mon].h - mons[mon].reserve_top - mons[mon].reserve_bottom - 2 * single_vertical_gap;
+
+			XWindowChanges wc = {
+				.x = x,
+				.y = y,
+				.width = MAX(1, w - 2 * border_width),
+				.height = MAX(1, h - 2 * border_width),
+				.border_width = border_width
+			};
+			XConfigureWindow(dpy, c->win,
+				CWX | CWY | CWWidth | CWHeight | CWBorderWidth, &wc);
+
+			c->x = wc.x;
+			c->y = wc.y;
+			c->w = wc.width;
+			c->h = wc.height;
+		}
+		update_borders();
+		return;
+	}
+
 	if (monocle) {
 		for (Client *c = head; c; c = c->next) {
 			if (!c->mapped || c->floating || c->fullscreen)
